app/src/main/java/ingsw/group1/findmyphone/contacts/SMSContactManager.java

```
@@ -2,45 +2,84 @@
 
 import android.content.Context;
 
+import androidx.annotation.NonNull;
 import androidx.annotation.Nullable;
 import androidx.room.Room;
 
-
 import com.eis.smslibrary.SMSPeer;
+import com.eis.smslibrary.exceptions.InvalidTelephoneNumberException;
 
 import java.util.List;
 
-
-
 /**
  * Class that takes care of inserting and deleting contacts from the database.
  * We don't use a support structure, like a Map,
  * because it is assumed that few contacts are inserted and not frequently.
- * So they can be added and deleted directly from the database every time.
+ * So they can be added and deleted directly from the {@link SMSContactDatabase} every time.
+ * It has been used design pattern Singleton.
  *
  * @author Giorgia Bortoletti
  */
-public class SMSContactManager {
+public class SMSContactManager implements ContactManager<String, SMSPeer, SMSContact>{
 
-    public static final String CONTACTS_DB_NAME = "contact-db";
+    private static final String CONTACTS_DB_NAME = "contact-db";
 
     private SMSContactDatabase contactDatabase;
+    private static SMSContactManager instance;
 
     //---------------------------- CONSTRUCTOR ----------------------------
 
     /**
-     * Constructor
+     * Private constructor
      *
      * @param applicationContext {@link Context} of the application
+     *
+     * @throws ExceptionInInitializerError if the object is already initialized
      */
-    public SMSContactManager(Context applicationContext) {
+    private SMSContactManager(@NonNull Context applicationContext) throws ExceptionInInitializerError {
+        if(instance != null)
```

Do you need to throw the exception if the constructor is private?

---

Maybe not but it might be useful for specs or not (?)

---

Technically you can break the private lock and access the constructor anyway, altough highly unlikely, I used to put this in my singletons